### DISCUSSION TITLE
how to add more plugin gstreamer  into acp docker

### DISCUSSION THREAD

**Discussion by vuhailongkl97** [AxisCommunications/discussions#168] [ACAP] [answered]
I want to use gstreamer for development on axis cameara.
Example: I want to run a pipeline like "rtspsrc ! rtph264depay ! fakesink" on a axis camera device but the process doesn't work due to missing rtspsrc plugin, ...
Can you help me add more gstreamer plugins on axis camera ?

**Comment by mattias-kindborg-at-work** [comment]
Hi @vuhailongkl97! Can you elaborate more on the use case? Are you aiming to do something special in the pipeline?

**Reply by vuhailongkl97** [reply]
@mattias-kindborg-at-work  
Currently I want to have some gst-plugins as original version of gst how to build and use it via installing app on device's web
I need libgstrtspclientsink but it wasn't existed in the device. how to i build and use it ? 
Thanks

**Comment by thanhlnbka** [comment]
@mattias-kindborg-at-work  I also want to use the gstream plugin to create another stream.
That stream will be pushed to the server by me and used for the purposes of our application
For example I would use a pipeline like this: "vdostream ! decodebin ! videoconvert ! x264enc ! rtspclientsink location=rtsp://...." or "rtspsrc ! rtph264depay ! rtspclientsink rtsp://...." .
Can you help me add more plugin on axis ? thanks

**Comment by daniel-falk** [comment, marked as answer]
@vuhailongkl97 and @thanhlnbka We are building our own gstreamer plugins for the cameras using [the official build instructions](https://gstreamer.freedesktop.org/documentation/installing/building-from-source-using-meson.html?gi-language=c). Most of it is straight forward (but a lot of work), but there are a few things to think of:

* If you are building plugins that you use together with gstreamer core already in the Axis cameras, then you need to build exactly the same version of gstreamer and you need to disable the debug log when building. This get a bit more complex since some of the cameras are using a gstreamer version from before gstreamer transitioned into a mono-repo, and newer cameras uses a version that needs to be built in the mono-repo. If you don't need to use the Axis specific plugins, then we find it easier to build the core too and use it as a stand alone library. In this case you need to spend some time optimizing the size of it to make it fit in the cameras flash (especially on low-end cameras like the S2L SoC).
* Some plugins like e.g. `rtspsrc` will forcefully override the maximum size of the socket buffers (as set in `rmem_max`). This can not be done as the `sdk` user as it does not have the `NET_ADMIN` capability, so you need to run the application as `root`, which will likely cause issues in the future.

Here is an example of the cross file needed to build for ARTPEC-8 cameras:
```ini
[host_machine]
system = 'linux'
cpu_family = 'aarch64'
cpu = 'aarch64'
endian = 'little'

[binaries]
pkgconfig = '/usr/bin/pkg-config'
```

**Reply by vuhailongkl97** [reply]
@daniel-falk thanks for your suggestion.
